src/dm/DispatchManagerActions.cc

```diff
@@ -2249,7 +2303,750 @@ int DispatchManager::live_updateconf(std::unique_ptr<VirtualMachine> vm,
+            error_str = oss.str();
+
+            return -1;
+        }
+
+        // ----------------------------------------------------------------------
+        // Update VM data
+        // - NIC to include new SG
+        // - Template to include new SG rules (note can be there already)
+        // ----------------------------------------------------------------------
+        vector<VectorAttribute *> sg_rules;
+
+        sgpool->get_security_group_rules(-1, sgid, sg_rules);
+
+        nic->add_security_group(sgid);
+
+        vm->remove_security_group(sgid); //duplicates
+
+        vm->add_template_attribute(sg_rules);
+
+        vmpool->update(vm.get());
+
+        if ( is_update )
+        {
+            vmm->updatesg(vm.get(), sgid);
+        }
+    }
+    else
+    {
+        oss << "Could not attach SG to VM " << vid << ", VM does not exist";
+        error_str = oss.str();
+
+        return -1;
+    }
+
+    if (auto sg = sgpool->get(sgid))
+    {
+        sg->del_vm(vid);
+
+        if ( is_update )
+        {
+            sg->add_updating(vid);
+        }
+        else
+        {
+            sg->add_vm(vid);
+        }
+
+        sgpool->update(sg.get());
+    }
+
+    return 0;
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+int DispatchManager::detach_sg(int vid, int nicid, int sgid,
+                               const RequestAttributes& ra, string& error_str)
+{
+    ostringstream oss;
+
+    bool is_update = false;
+
+    vector<VectorAttribute *> sg_rules;
+
+    // -------------------------------------------------------------------------
+    // Check action consistency:
+    // - Object exists VM, NIC and SG
+    // - State is compatible with operation
+    // -------------------------------------------------------------------------
+    if ( auto vm = vmpool->get(vid) )
+    {
+        VirtualMachine::LcmState lstate = vm->get_lcm_state();
+
+        switch (lstate)
+        {
+            //Cannnot update VM, SG rules being updated/created
+            case VirtualMachine::BOOT:
+            case VirtualMachine::BOOT_MIGRATE:
+            case VirtualMachine::BOOT_SUSPENDED:
+            case VirtualMachine::BOOT_STOPPED:
+            case VirtualMachine::BOOT_UNDEPLOY:
+            case VirtualMachine::BOOT_POWEROFF:
+            case VirtualMachine::BOOT_UNKNOWN:
+            case VirtualMachine::BOOT_FAILURE:
+            case VirtualMachine::BOOT_MIGRATE_FAILURE:
+            case VirtualMachine::BOOT_UNDEPLOY_FAILURE:
+            case VirtualMachine::BOOT_STOPPED_FAILURE:
+            case VirtualMachine::MIGRATE:
+            case VirtualMachine::HOTPLUG_NIC:
+            case VirtualMachine::HOTPLUG_NIC_POWEROFF:
+            case VirtualMachine::UNKNOWN:
+                oss << "VM " << vid << " is in wrong state " << vm->state_str();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Potential log injection detected. Ensure all untrusted input is properly sanitized before logging. Use parameterized logging or validate input against an allow list to prevent log injection vulnerabilities. Consider using a dedicated logging library&#39;s built-in sanitization features when available. Learn more - https://cwe.mitre.org/data/definitions/117.html